File: .agent/skills-backup-pre-migration/SelfEvolution/scripts/version_manager.py
#!/usr/bin/env python3
"""
Version Manager for SelfEvolution v4.0
Manages version history, changelog, and rollback capability
"""

import yaml
import json
import shutil
import sys
from pathlib import Path
from datetime import datetime
from typing import Dict, List, Optional

# Add scripts directory to path
SCRIPT_DIR = Path(__file__).parent
sys.path.insert(0, str(SCRIPT_DIR))

from project_utils import (
    get_mistakes_file,
    get_improvements_file,
    get_meta_file,
    get_versions_dir,
    ensure_v4_structure
)

def save_version(version_type: str, version_number: int, data: Dict) -> Path:
    """
    Save a version snapshot
    
    Args:
        version_type: 'mistakes' or 'improvements'
        version_number: Version number
        data: Data to save
        
    Returns:
        Path to saved version file
    """
    ensure_v4_structure()
    versions_dir = get_versions_dir()
    
    version_file = versions_dir / f'{version_type}-v{version_number}.yaml'
    
    with open(version_file, 'w', encoding='utf-8') as f:
        yaml.dump(data, f, allow_unicode=True, default_flow_style=False, sort_keys=False)
    
    return version_file

def load_version(version_type: str, version_number: int) -> Optional[Dict]:
    """
    Load a specific version
    
    Args:
        version_type: 'mistakes' or 'improvements'
        version_number: Version number to load
        
    Returns:
        Version data or None if not found
    """
    versions_dir = get_versions_dir()
    version_file = versions_dir / f'{version_type}-v{version_number}.yaml'
    
    if not version_file.exists():
        return None
    
    with open(version_file, 'r', encoding='utf-8') as f:
        return yaml.safe_load(f)

def list_versions(version_type: str) -> List[Dict]:
    """
    List all versions of a type
    
    Args:
        version_type: 'mistakes' or 'improvements'
        
    Returns:
        List of version info dicts
    """
    versions_dir = get_versions_dir()
    
    if not versions_dir.exists():
        return []
    
    version_files = sorted(versions_dir.glob(f'{version_type}-v*.yaml'))
    
    versions = []
    for v_file in version_files:
        # Extract version number
        version_str = v_file.stem.split('-v')[1]
        version_num = int(version_str)
        
        # Get file stats
        stat = v_file.stat()
        
        # Load data to get metadata
        with open(v_file, 'r', encoding='utf-8') as f:
            data = yaml.safe_load(f) or {}
        
        count = len(data.get(version_type, []))
        last_improved = data.get('last_improved')
        
        versions.append({
            'version': version_num,
            'file': str(v_file),
            'size': stat.st_size,
            'modified': datetime.fromtimestamp(stat.st_mtime).isoformat(),
            'count': count,
            'last_improved': last_improved
        })
    
    return versions

def get_current_version(version_type: str) -> int:
    """
    Get current version number
    
    Args:
        version_type: 'mistakes' or 'improvements'
        
    Returns:
        Current version number
    """
    if version_type == 'mistakes':
        file_path = get_mistakes_file()
    else:
        file_path = get_improvements_file()
    
    if not file_path.exists():
        return 0
    
    with open(file_path, 'r', encoding='utf-8') as f:
        data = yaml.safe_load(f) or {}
    
    return data.get('version', 1)

def rollback_to_version(version_type: str, target_version: int) -> bool:
    """
    Rollback to a specific version
    
    Args:
        version_type: 'mistakes' or 'improvements'  
        target_version: Version number to rollback to
        
    Returns:
        True if successful, False otherwise
    """
    # Load target version
    target_data = load_version(version_type, target_version)
    
    if not target_data:
        print(f"❌ Version {target_version} not found")
        return False
    
    # Get current file
    if version_type == 'mistakes':
        current_file = get_mistakes_file()
    else:
        current_file = get_improvements_file()
    
    # Backup current version first
    current_version = get_current_version(version_type)
    backup_file = current_file.parent / f'{version_type}.backup-v{current_version}.yaml'
    
    if current_file.exists():
        shutil.copy2(current_file, backup_file)
        print(f"💾 Backed up current version to {backup_file.name}")
    
    # Restore target version
    with open(current_file, 'w', encoding='utf-8') as f:
        yaml.dump(target_data, f, allow_unicode=True, default_flow_style=False, sort_keys=False)
    
    print(f"✅ Rolled back to version {target_version}")
    
    # Update meta.json
    try:
        meta_file = get_meta_file()
        if meta_file.exists():
            with open(meta_file, 'r', encoding='utf-8') as f:
                meta = json.load(f)
            
            # Add rollback entry to history
            if 'rollbacks' not in meta:
                meta['rollbacks'] = []
            
            meta['rollbacks'].append({
                'type': version_type,
                'from_version': current_version,
                'to_version': target_version,
                'rolled_back_at': datetime.now().isoformat()
            })
            
            meta['updated'] = datetime.now().isoformat()
            
            with open(meta_file, 'w', encoding='utf-8') as f:
                json.dump(meta, f, indent=2, ensure_ascii=False)
    except Exception as e:
        print(f"⚠️  Could not update meta.json: {e}")
    
    return True

def compare_versions(version_type: str, version_a: int, version_b: int) -> Dict:
    """
    Compare two versions
    
    Args:
        version_type: 'mistakes' or 'improvements'
        version_a: First version
        version_b: Second version
        
    Returns:
        Dict with comparison results
    """
    data_a = load_version(version_type, version_a)
    data_b = load_version(version_type, version_b)
    
    if not data_a or not data_b:
        return {'error': 'One or both versions not found'}
    
    items_a = data_a.get(version_type, [])
    items_b = data_b.get(version_type, [])
    
    # Get IDs
    ids_a = set(item['id'] for item in items_a)
    ids_b = set(item['id'] for item in items_b)
    
    # Find differences
    added = ids_b - ids_a
    removed = ids_a - ids_b
    common = ids_a & ids_b
    
    # Check for modifications in common items
    modified = []
    for item_id in common:
        item_a = next(i for i in items_a if i['id'] == item_id)
        item_b = next(i for i in items_b if i['id'] == item_id)
        
        if item_a.get('version', 1) != item_b.get('version', 1):
            modified.append({
                'id': item_id,
                'version_a': item_a.get('version', 1),
                'version_b': item_b.get('version', 1)
            })
    
    return {
        'version_a': version_a,
        'version_b': version_b,
        'total_a': len(items_a),
        'total_b': len(items_b),
        'added': list(added),
        'removed': list(removed),
        'modified': modified
    }

def display_changelog(version_type: str, version_number: int):
    """
    Display changelog for a version
    
    Args:
        version_type: 'mistakes' or 'improvements'
        version_number: Version to display
    """
    data = load_version(version_type, version_number)
    
    if not data:
        print(f"Version {version_number} not found")
        return
    
    items = data.get(version_type, [])
    
    print(f"\n📋 Changelog for {version_type} v{version_number}")
    print(f"Last improved: {data.get('last_improved', 'Never')}")
    print(f"Total items: {len(items)}\n")
    
    for item in items:
        changelog = item.get('changelog', [])
        if not changelog:
            continue
        
        print(f"{item['id']} (v{item.get('version', 1)}):")
        for entry in changelog:
            print(f"  v{entry.get('version', '?')}: {entry.get('change', 'No details')}")
            print(f"     Date: {entry.get('date', 'Unknown')[:10]}")
        print()

def main():
    """CLI entry point"""
    import argparse
    
    parser = argparse.ArgumentParser(description='Version Manager for SelfEvolution v4.0')
    subparsers = parser.add_subparsers(dest='command', help='Commands')
    
    # List versions
    list_parser = subparsers.add_parser('list', help='List all versions')
    list_parser.add_argument('type', choices=['mistakes', 'improvements'], help='Type')
    
    # Show version details
    show_parser = subparsers.add_parser('show', help='Show version details')
    show_parser.add_argument('type', choices=['mistakes', 'improvements'], help='Type')
    show_parser.add_argument('version', type=int, help='Version number')
    
    # Rollback
    rollback_parser = subparsers.add_parser('rollback', help='Rollback to version')
    rollback_parser.add_argument('type', choices=['mistakes', 'improvements'], help='Type')
    rollback_parser.add_argument('version', type=int, help='Target version')
    
    # Compare
    compare_parser = subparsers.add_parser('compare', help='Compare versions')
    compare_parser.add_argument('type', choices=['mistakes', 'improvements'], help='Type')
    compare_parser.add_argument('version_a', type=int, help='First version')
    compare_parser.add_argument('version_b', type=int, help='Second version')
    
    # Changelog
    changelog_parser = subparsers.add_parser('changelog', help='Show changelog')
    changelog_parser.add_argument('type', choices=['mistakes', 'improvements'], help='Type')
    changelog_parser.add_argument('version', type=int, help='Version number')
    
    args = parser.parse_args()
    
    if args.command == 'list':
        versions = list_versions(args.type)
        
        if not versions:
            print(f"No versions found for {args.type}")
            return
        
        print(f"\n📚 {args.type.capitalize()} Versions:\n")
        for v in versions:
            print(f"v{v['version']}: {v['modified'][:10]} - {v['count']} items ({v['size']} bytes)")
    
    elif args.command == 'show':
        data = load_version(args.type, args.version)
        
        if not data:
            print(f"Version {args.version} not found")
            return
        
        print(f"\n{args.type.capitalize()} v{args.version}:")
        print(f"Last improved: {data.get('last_improved', 'Never')}")
        print(f"Total items: {len(data.get(args.type, []))}")
        
        print(f"\nItems:")
        for item in data.get(args.type, []):
            print(f"  {item['id']} (v{item.get('version', 1)}): {item.get('lesson', item.get('improvement', ''))[:60]}")
    
    elif args.command == 'rollback':
        confirm = input(f"⚠️  Rollback {args.type} to v{args.version}? (yes/no): ")
        
        if confirm.lower() == 'yes':
            rollback_to_version(args.type, args.version)
        else:
            print("Cancelled")
    
    elif args.command == 'compare':
        result = compare_versions(args.type, args.version_a, args.version_b)
        
        if 'error' in result:
            print(f"❌ {result['error']}")
            return
        
        print(f"\n📊 Comparing v{args.version_a} vs v{args.version_b}:")
        print(f"  v{args.version_a}: {result['total_a']} items")
        print(f"  v{args.version_b}: {result['total_b']} items")
        print(f"\n  Added: {len(result['added'])} - {result['added']}")
        print(f"  Removed: {len(result['removed'])} - {result['removed']}")
        print(f"  Modified: {len(result['modified'])}")
        
        for mod in result['modified']:
            print(f"    {mod['id']}: v{mod['version_a']} → v{mod['version_b']}")
    
    elif args.command == 'changelog':
        display_changelog(args.type, args.version)
    
    else:
        parser.print_help()

if __name__ == "__main__":
    main()

File: .agents/skills/self-evolution/scripts/version_manager.js
#!/usr/bin/env node
/**
 * Version Manager for SelfEvolution v4.0
 * Manages version history, changelog, and rollback capability
 */

import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';
import { fileURLToPath } from 'url';
import readline from 'readline';

import {
    getMistakesFile,
    getImprovementsFile,
    getMetaFile,
    getVersionsDir,
    ensureV4Structure
} from './project_utils.js';

const __filename = fileURLToPath(import.meta.url);

/**
 * Save a version snapshot
 */
export function saveVersion(versionType, versionNumber, data) {
    ensureV4Structure();
    const versionsDir = getVersionsDir();

    fs.mkdirSync(versionsDir, { recursive: true });
    const versionFile = path.join(versionsDir, `${versionType}-v${versionNumber}.yaml`);

    fs.writeFileSync(versionFile, yaml.dump(data, { noRefs: true }), 'utf-8');

    return versionFile;
}

/**
 * Load a specific version
 */
export function loadVersion(versionType, versionNumber) {
    const versionsDir = getVersionsDir();
    const versionFile = path.join(versionsDir, `${versionType}-v${versionNumber}.yaml`);

    if (!fs.existsSync(versionFile)) {
        return null;
    }

    return yaml.load(fs.readFileSync(versionFile, 'utf-8'));
}

/**
 * List all versions of a type
 */
export function listVersions(versionType) {
    const versionsDir = getVersionsDir();

    if (!fs.existsSync(versionsDir)) {
        return [];
    }

    const files = fs.readdirSync(versionsDir)
        .filter(f => f.startsWith(`${versionType}-v`) && f.endsWith('.yaml'))
        .sort();

    const versions = [];
    for (const vFile of files) {
        const versionStr = vFile.split('-v')[1].replace('.yaml', '');
        const versionNum = parseInt(versionStr, 10);

        const filePath = path.join(versionsDir, vFile);
        const stat = fs.statSync(filePath);

        const data = yaml.load(fs.readFileSync(filePath, 'utf-8')) || {};
        const count = (data[versionType] || []).length;
        const lastImproved = data.last_improved;

        versions.push({
            version: versionNum,
            file: filePath,
            size: stat.size,
            modified: new Date(stat.mtime).toISOString(),
            count,
            last_improved: lastImproved
        });
    }

    return versions;
}

/**
 * Get current version number
 */
export function getCurrentVersion(versionType) {
    const filePath = versionType === 'mistakes'
        ? getMistakesFile()
        : getImprovementsFile();

    if (!fs.existsSync(filePath)) {
        return 0;
    }

    const data = yaml.load(fs.readFileSync(filePath, 'utf-8')) || {};
    return data.version || 1;
}

/**
 * Rollback to a specific version
 */
export function rollbackToVersion(versionType, targetVersion) {
    const targetData = loadVersion(versionType, targetVersion);

    if (!targetData) {
        console.log(`❌ Version ${targetVersion} not found`);
        return false;
    }

    const currentFile = versionType === 'mistakes'
        ? getMistakesFile()
        : getImprovementsFile();

    // Backup current version first
    const currentVersion = getCurrentVersion(versionType);
    const backupFile = path.join(path.dirname(currentFile), `${versionType}.backup-v${currentVersion}.yaml`);

    if (fs.existsSync(currentFile)) {
        fs.copyFileSync(currentFile, backupFile);
        console.log(`💾 Backed up current version to ${path.basename(backupFile)}`);
    }

    // Restore target version
    fs.writeFileSync(currentFile, yaml.dump(targetData, { noRefs: true }), 'utf-8');

    console.log(`✅ Rolled back to version ${targetVersion}`);

    // Update meta.json
    try {
        const metaFile = getMetaFile();
        if (fs.existsSync(metaFile)) {
            const meta = JSON.parse(fs.readFileSync(metaFile, 'utf-8'));

            if (!meta.rollbacks) meta.rollbacks = [];
            meta.rollbacks.push({
                type: versionType,
                from_version: currentVersion,
                to_version: targetVersion,
                rolled_back_at: new Date().toISOString()
            });

            meta.updated = new Date().toISOString();

            fs.writeFileSync(metaFile, JSON.stringify(meta, null, 2), 'utf-8');
        }
    } catch (e) {
        console.log(`⚠️  Could not update meta.json: ${e.message}`);
    }

    return true;
}

/**
 * Compare two versions
 */
export function compareVersions(versionType, versionA, versionB) {
    const dataA = loadVersion(versionType, versionA);
    const dataB = loadVersion(versionType, versionB);

    if (!dataA || !dataB) {
        return { error: 'One or both versions not found' };
    }

    const itemsA = dataA[versionType] || [];
    const itemsB = dataB[versionType] || [];

    const idsA = new Set(itemsA.map(item => item.id));
    const idsB = new Set(itemsB.map(item => item.id));

    const added = [...idsB].filter(id => !idsA.has(id));
    const removed = [...idsA].filter(id => !idsB.has(id));
    const common = [...idsA].filter(id => idsB.has(id));

    const modified = [];
    for (const itemId of common) {
        const itemA = itemsA.find(i => i.id === itemId);
        const itemB = itemsB.find(i => i.id === itemId);

        if ((itemA.version || 1) !== (itemB.version || 1)) {
            modified.push({
                id: itemId,
                version_a: itemA.version || 1,
                version_b: itemB.version || 1
            });
        }
    }

    return {
        version_a: versionA,
        version_b: versionB,
        total_a: itemsA.length,
        total_b: itemsB.length,
        added,
        removed,
        modified
    };
}

/**
 * Display changelog for a version
 */
export function displayChangelog(versionType, versionNumber) {
    const data = loadVersion(versionType, versionNumber);

    if (!data) {
        console.log(`Version ${versionNumber} not found`);
        return;
    }

    const items = data[versionType] || [];

    console.log(`\n📋 Changelog for ${versionType} v${versionNumber}`);
    console.log(`Last improved: ${data.last_improved || 'Never'}`);
    console.log(`Total items: ${items.length}\n`);

    for (const item of items) {
        const changelog = item.changelog || [];
        if (changelog.length === 0) continue;

        console.log(`${item.id} (v${item.version || 1}):`);
        for (const entry of changelog) {
            console.log(`  v${entry.version || '?'}: ${entry.change || 'No details'}`);
            console.log(`     Date: ${(entry.date || 'Unknown').slice(0, 10)}`);
        }
        console.log();
    }
}

function prompt(question) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    return new Promise(resolve => {
        rl.question(question, answer => {
            rl.close();
            resolve(answer);
        });
    });
}

async function main() {
    const args = process.argv.slice(2);
    const command = args[0];

    switch (command) {
        case 'list': {
            const versionType = args[1];
            if (!['mistakes', 'improvements'].includes(versionType)) {
                console.log('Usage: node version_manager.js list <mistakes|improvements>');
                return;
            }

            const versions = listVersions(versionType);

            if (versions.length === 0) {
                console.log(`No versions found for ${versionType}`);
                return;
            }

            console.log(`\n📚 ${versionType.charAt(0).toUpperCase() + versionType.slice(1)} Versions:\n`);
            for (const v of versions) {
                console.log(`v${v.version}: ${v.modified.slice(0, 10)} - ${v.count} items (${v.size} bytes)`);
            }
            break;
        }

        case 'show': {
            const versionType = args[1];
            const versionNum = parseInt(args[2], 10);

            const data = loadVersion(versionType, versionNum);

            if (!data) {
                console.log(`Version ${versionNum} not found`);
                return;
            }

            console.log(`\n${versionType.charAt(0).toUpperCase() + versionType.slice(1)} v${versionNum}:`);
            console.log(`Last improved: ${data.last_improved || 'Never'}`);
            console.log(`Total items: ${(data[versionType] || []).length}`);

            console.log('\nItems:');
            for (const item of data[versionType] || []) {
                const text = (item.lesson || item.improvement || '').slice(0, 60);
                console.log(`  ${item.id} (v${item.version || 1}): ${text}`);
            }
            break;
        }

        case 'rollback': {
            const versionType = args[1];
            const targetVersion = parseInt(args[2], 10);

            const confirm = await prompt(`⚠️  Rollback ${versionType} to v${targetVersion}? (yes/no): `);

            if (confirm.toLowerCase() === 'yes') {
                rollbackToVersion(versionType, targetVersion);
            } else {
                console.log('Cancelled');
            }
            break;
        }

        case 'compare': {
            const versionType = args[1];
            const versionA = parseInt(args[2], 10);
            const versionB = parseInt(args[3], 10);

            const result = compareVersions(versionType, versionA, versionB);

            if (result.error) {
                console.log(`❌ ${result.error}`);
                return;
            }

            console.log(`\n📊 Comparing v${versionA} vs v${versionB}:`);
            console.log(`  v${versionA}: ${result.total_a} items`);
            console.log(`  v${versionB}: ${result.total_b} items`);
            console.log(`\n  Added: ${result.added.length} - ${JSON.stringify(result.added)}`);
            console.log(`  Removed: ${result.removed.length} - ${JSON.stringify(result.removed)}`);
            console.log(`  Modified: ${result.modified.length}`);

            for (const mod of result.modified) {
                console.log(`    ${mod.id}: v${mod.version_a} → v${mod.version_b}`);
            }
            break;
        }

        case 'changelog': {
            const versionType = args[1];
            const versionNum = parseInt(args[2], 10);
            displayChangelog(versionType, versionNum);
            break;
        }

        default:
            console.log('Version Manager for SelfEvolution v4.0');
            console.log('\nCommands:');
            console.log('  list <type>                   List all versions');
            console.log('  show <type> <version>         Show version details');
            console.log('  rollback <type> <version>     Rollback to version');
            console.log('  compare <type> <v1> <v2>      Compare versions');
            console.log('  changelog <type> <version>    Show changelog');
            console.log('\nTypes: mistakes, improvements');
    }
}

if (process.argv[1] === __filename) {
    main();
}
